pkg/daemon/server/service/pipeline_watermark_query.go

```diff
@@ -26,46 +26,48 @@ import (
 	"github.com/numaproj/numaflow/pkg/apis/proto/daemon"
 	"github.com/numaproj/numaflow/pkg/isbsvc"
 	"github.com/numaproj/numaflow/pkg/watermark/fetch"
-	"github.com/numaproj/numaflow/pkg/watermark/processor"
+	"github.com/numaproj/numaflow/pkg/watermark/store"
 )
 
 // GetUXEdgeWatermarkFetchers returns a map of the watermark fetchers, where key is the buffer name,
 // value is a list of fetchers to the buffers.
-func GetUXEdgeWatermarkFetchers(ctx context.Context, pipeline *v1alpha1.Pipeline, processorManagers map[v1alpha1.Edge][]*processor.ProcessorManager) (map[v1alpha1.Edge][]fetch.UXFetcher, error) {
+func GetUXEdgeWatermarkFetchers(ctx context.Context, pipeline *v1alpha1.Pipeline, wmStores map[v1alpha1.Edge][]store.WatermarkStore) (map[v1alpha1.Edge][]fetch.UXFetcher, error) {
 	var wmFetchers = make(map[v1alpha1.Edge][]fetch.UXFetcher)
 	if pipeline.Spec.Watermark.Disabled {
 		return wmFetchers, nil
 	}
 
-	for edge, pms := range processorManagers {
+	for edge, stores := range wmStores {
 		var fetchers []fetch.UXFetcher
-		for _, pm := range pms {
-			fetchers = append(fetchers, fetch.NewEdgeFetcher(ctx, pm, pipeline.GetVertex(edge.To).GetPartitionCount()))
+		isReduce := pipeline.GetVertex(edge.To).IsReduceUDF()
+		partitionCount := pipeline.GetVertex(edge.To).GetPartitionCount()
+		for i, s := range stores {
+			fetchers = append(fetchers, fetch.NewEdgeFetcher(ctx, s, partitionCount, fetch.WithIsReduce(isReduce), fetch.WithVertexReplica(int32(i))))
 		}
 		wmFetchers[edge] = fetchers
 	}
 
 	return wmFetchers, nil
 }
 
-// GetProcessorManagers returns a map of ProcessorManager per edge.
-func GetProcessorManagers(ctx context.Context, pipeline *v1alpha1.Pipeline, isbsvcClient isbsvc.ISBService) (map[v1alpha1.Edge][]*processor.ProcessorManager, error) {
-	var processorManagers = make(map[v1alpha1.Edge][]*processor.ProcessorManager)
+// GetWatermarkStores returns a map of watermark stores per edge.
+func GetWatermarkStores(ctx context.Context, pipeline *v1alpha1.Pipeline, isbsvcClient isbsvc.ISBService) (map[v1alpha1.Edge][]store.WatermarkStore, error) {
```

Comment:
nit: Would `CreateWatermarkStores` be a better name? The word `Get` kind of indicates that the stores are already there and the function is just returning a reference to them, although the function is actually Newing.

Same for `GetUXEdgeWatermarkFetchers`